lib/rules/id-match.js

```diff
@@ -106,36 +143,70 @@ module.exports = {
                             report(node);
                         }
 
-                    // Report AssignmentExpressions only if they are the left side of the assignment
+                    // Report AssignmentExpressions left side's assigned variable id
                     } else if (effectiveParent.type === "AssignmentExpression" &&
-                        (effectiveParent.right.type !== "MemberExpression" ||
                         effectiveParent.left.type === "MemberExpression" &&
-                        effectiveParent.left.property.name === name)) {
+                        effectiveParent.left.property.name === node.name) {
+                        if (isInvalid(name)) {
+                            report(node);
+                        }
+
+                    // Report AssignmentExpressions only if they are the left side of the assignment
+                    } else if (effectiveParent.type === "AssignmentExpression" && effectiveParent.right.type !== "MemberExpression") {
                         if (isInvalid(name)) {
                             report(node);
                         }
                     }
 
-                } else if (parent.type === "Property") {
+                /*
+                 * Properties have their own rules, and
+                 * AssignmentPattern nodes can be treated like Properties:
+                 * e.g.: const { no_camelcased = false } = bar;
+                 */
+                } else if (parent.type === "Property" || node.parent.type === "AssignmentPattern") {
```

Comment:
> node.parent.type === "AssignmentPattern"

nits: is the `parent` means `node.parent`? I'd slightly prefer using `parent` for consistency.

Reply:
Yes, I'll make a change.